### PR TITLE
Introduce addAlias method.

### LINKED ISSUE #824: Libzim should support alias creation.
For https://github.com/kiwix/overview/issues/95

We need to be able to create entries pointing to the same data.

### PR DESCRIPTION
**Pull request by mgautierfr**
This is the first PR for https://github.com/kiwix/overview/issues/95

Fix #824

### CODE REVIEW

**Review by veloman-yunkan — COMMENTED** [review]
In the current implementation information about presence of clones is nowhere recorded.

Have conscious decisions been made (and properly implemented) regarding the following questions:

1. How should clones be accounted for in media type counts?
2. How should full text indexing on cloned entries work? If FT indexing associates a document only with the entry added via the `zim::Creator::addItem()` method, then the cloned dirents better be named aliases because of the resulting asymmetry.
3. Shouldn't it be possible to iterate over items while skipping cloned entries (i.e. having each data item returned only once)?
4. Shouldn't it be possible to find out all the clone-siblings of a given dirent?

**Comment by mgautierfr** [comment]
> In the current implementation information about presence of clones in nowhere recorded.

It is somehow intended. In fact nothing in the specification (from the beginning) prevent two dirent to point to the same data. It is just that no implementation allow that.
I don't see why we should inform about their presence. Especially, reader should always be prepared to that.

Have conscious decisions been made (and properly implemented) regarding the following questions:

> - How should clones be accounted for in media type counts?
> - How should full text indexing on cloned entries work? If FT indexing associates a document only with the entry added via the `zim::Creator::addItem()` method, then the cloned dirents better be named aliases because of the resulting asymmetry.

Indeed. For now, clones are handled "as redirect":
- They are not counted in the media type counter
- They are not fulltext indexed. (Only title)

This is mainly a technical limitation as we don't have access to the content (only the cluster/blob index) when we add the clone.

> Shouldn't it be possible to iterate over items while skipping cloned entries (i.e. having each data item returned only once)?
> Shouldn't it be possible to find out all the clone-siblings of a given dirent?

I'm not sure we need that. The only need I see for now is in zimcheck but I handle that there.


---

This new feature may evolve and be used for something not foreseen. But for now, the use case is to "reference" binary content (as youtube video) under different names for zimit files.


> [...] then the cloned dirents better be named aliases because of the resulting asymmetry.

This a good point. I agree with you. Maybe alias is a better name.

I have named it with a generic name as it is technically a generic feature but the intended use case is pretty limited:
- We will not clone articles (so no problem about indexation)
- This is used to resolve dynamically generated url, so the content must be counted for one.

So I don't think we need to do more, at least for now. If other use cases appear, we may extend our API accordingly.

File: include/zim/writer/creator.h [hunk 1]
```diff
@@ -195,6 +195,36 @@ namespace zim
...
+         * The existing entry pointed by `targetPath` is cloned and updated with
+         * `path` and `title`.
+         *
+         * The alias entry will shared the same type (redirection or item)
+         * and namespace than `targetPath`.
+         *
+         * If the `targetPath` is a item, the new entry will be item pointing
+         * to the same data than `targetPath` item. (Not a redirection to `targetPath`).
+         * However, the alias entry is not counted in the media type counter
+         * and it is not fulltext indexed (only title indexed).
```
Comment (veloman-yunkan) [inline include/zim/writer/creator.h:211]: This indeed seems to be the case, but it better be demonstrated by tests.

File: src/writer/direntPool.h [hunk 2]
```diff
@@ -65,22 +73,22 @@ namespace zim
...
-          if (direntIndex == 0xFFFF) {
-            allocate_new_pool();
-          }
-          auto dirent = pools.back() + direntIndex++;
+          auto dirent = getDirentSlot();
           new (dirent) Dirent(ns, path, title, targetNs, targetPath);
           return dirent;
         }
+
+        Dirent* getCloneDirent(const std::string& path, const std::string& title, const Dirent& target) {
```
Comment (veloman-yunkan) [inline src/writer/direntPool.h:87]: Did this one escape the `Clone` → `Alias` renaming or it is intentional?

**Comment by veloman-yunkan** [comment]
> > Shouldn't it be possible to iterate over items while skipping cloned entries (i.e. having each data item returned only once)?
> > Shouldn't it be possible to find out all the clone-siblings of a given dirent?
> 
> I'm not sure we need that. The only need I see for now is in zimcheck but I handle that there.

Recovering information about a dirent being an alias requires extra effort. Can't we preserve it easily? For example is [`Dirent::parameter`](https://github.com/openzim/libzim/blob/2a810acaddb9dc85224aa5222209fb92128f0159/src/_dirent.h#L52) used for anything? If not, why don't we record the dirent's purpose there?

**Comment by mgautierfr** [comment]
> [Dirent::parameter](https://github.com/openzim/libzim/blob/2a810acaddb9dc85224aa5222209fb92128f0159/src/_dirent.h#L52) used for anything? If not, why don't we record the dirent's purpose there?

Dirent::parameter is indeed not used for now. But is our only way to extend a dirent without changing the zim format so if we start to use it, we have to define a way to use it for several things. If we use it "blindly" to store "isClone" we lost a way to store other values.

The idea of this change was it was "creator" only and do not change anything on reading part.
Let's me think on this 🙂

**Comment by kelson42** [comment]
@mgautierfr Status not clear to me, seems you are over your side but new review has not been requested.

File: src/writer/direntPool.h (hunk 2 above)
Comment (mgautierfr) [reply src/writer/direntPool.h:87]: It was, as at this level, we are creating a Clone.
The difference in behavior (indexing) is done as a higher level.

But keeping the "clone" semantic only here will mostly disturb us.

**Comment by mgautierfr** [comment]
> Status not clear to me, seems you are over your side but new review has not been requested.

PR updated and ready for another review.

What is left open is the question about extending the format to use `Dirent::parameter`. 
I still not convince about that. But neither have a strong opinion on the opposite side.

I would suggest that we go with this PR and discuss the extension of the format in a issue.
But the issue would have to be fixed/resolved BEFORE we do another release and start creating zim file with alias.

As long as I have not finished https://github.com/kiwix/overview/issues/95, there is no need to merge it. So maybe we can simply continue the review process to agree with @veloman-yunkan.

**Comment by kelson42** [comment]
@veloman-yunkan 

> What is left open is the question about extending the format to use `Dirent::parameter`.
> I still not convince about that. But neither have a strong opinion on the opposite side.

I'm not convince about the necessity to do that in general, in right now in particular, I would recommend to open a dedicated ticket to keep track of this idea.

That said, we should be able to identify clearly a ZIM file which clearly potentialy have such clones (even if this was allowed implicitly in the past, it will be from now explicitly) I would recommend to:
* increment the ZIM version number by a minor version
* Update the spec to clearly say it is allowed

**Review by veloman-yunkan — COMMENTED** [review]
LGTM. Fixup commits must be eliminated.

**Pull request merged**